pilot/pkg/model/push_context.go

```diff
@@ -1809,3 +1810,82 @@ func (ps *PushContext) ServiceInstancesByPort(svc *Service, port int, labels lab
 	// Fallback to discovery call.
 	return ps.InstancesByPort(svc, port, labels)
 }
+
+// Maybe rewrite gateway server port from target port to service port if:
+// * GW server port is unambiguously a target port (i.e. there is no service port using the same port number)
+// * There is only one service port mapped to this target port, which the gateway server port will be rewritten to.
+// * There is only one target port mapped to the rewritten service port (i.e. there is no ambiguity later on when generating listener with target port)
+// * Another gateway server is defined to listen on the service port, thus without rewriting, duplicated listeners will be generated.
+func maybeRewriteGatewayPort(svcs []*ServiceInstance, gwc []config.Config) []config.Config {
```

Comment:
do we handle the VirtualService port matches? And the port match in the vhost matcher (it matches Host header of `foo:<gateway server port>`)?

Reply:
Seems like we don't (also not sure how we possibly could). The thing that is pretty weird about this though is that you may define a gateway + VS and it works fine, then another gateway is added (triggering the "Another gateway server is defined to listen on the service port") and suddenly it behaves different

Reply:
Thinking about it a bit, I think there is no perfect solution. It seems the API is flawed in that the `port` is ambiguous, and may refer to either targetPort translation or directly to a port; the change to do the targetPort mapping likely didn't consider these edge cases?

So while this is not ideal, maybe its the best option. Any opinions @ramaraochavali @hzxuzhonghu @costinm @nrjpoddar ?

Reply:
> another gateway is added (triggering the "Another gateway server is defined to listen on the service port") and suddenly it behaves different

If all the conditions are met, it will not only behave differently, but it will simply not work due to duplicated listeners.

Reply:
ah right.. good point, so its strictly better. Makes sense to me. i'll approve but lets wait a bit in case anyone else has opinions?